src/Storages/ObjectStorage/StorageObjectStorage.cpp

```diff
@@ -173,8 +175,80 @@ StorageObjectStorage::StorageObjectStorage(
         }
     }
 
-    setVirtuals(VirtualColumnUtils::getVirtualsForFileLikeStorage(
-        metadata.columns, context, sample_path, format_settings, configuration->isDataLakeConfiguration()));
+    /*
+     * If `partition_strategy=hive`, the partition columns shall be extracted from the `PARTITION BY` expression.
+     * There is no need to read from the filepath.
+     *
+     * Otherwise, in case `use_hive_partitioning=1`, we can keep the old behavior of extracting it from the sample path.
+     * And if the schema was inferred (not specified in the table definition), we need to enrich it with the path partition columns
+     */
+    if (configuration->partition_strategy && configuration->partition_strategy_type == PartitionStrategyFactory::StrategyType::HIVE)
+    {
+        hive_partition_columns_to_read_from_file_path = configuration->partition_strategy->getPartitionColumns();
+    }
+    else if (context->getSettingsRef()[Setting::use_hive_partitioning])
+    {
+        HivePartitioningUtils::extractPartitionColumnsFromPathAndEnrichStorageColumns(
+           columns,
+           hive_partition_columns_to_read_from_file_path,
+           sample_path,
+           columns_in_table_or_function_definition.empty(),
+           format_settings,
+           context);
+    }
+
+    if (hive_partition_columns_to_read_from_file_path.size() == columns.size())
+    {
+        throw Exception(
+            ErrorCodes::INCORRECT_DATA,
+            "A hive partitioned file can't contain only partition columns. Try reading it with `partition_strategy=wildcard` and `use_hive_partitioning=0`");
+    }
+
+    if (configuration->partition_columns_in_data_file)
+    {
+        file_columns = columns;
+    }
+    else
+    {
+        std::unordered_set<String> hive_partition_columns_to_read_from_file_path_set;
+
+        for (const auto & [name, type] : hive_partition_columns_to_read_from_file_path)
+        {
+            hive_partition_columns_to_read_from_file_path_set.insert(name);
+        }
+
+        for (const auto & [name, type] : columns.getAllPhysical())
+        {
+            if (!hive_partition_columns_to_read_from_file_path_set.contains(name))
+            {
+                file_columns.add({name, type});
+            }
+        }
+    }
+
+    // Assert file contains at least one column. The assertion only takes place if we were able to deduce the schema. The storage might be empty.
+    if (!columns.empty())
+    {
+        if (file_columns.empty())
+        {
+            throw Exception(ErrorCodes::INCORRECT_DATA,
+                "File without physical columns is not supported. Give it a try with `use_hive_partitioning=0` and or `partition_strategy=wildcard`. File {}",
```

Comment:
```suggestion
                "File without physical columns is not supported. Please try with `use_hive_partitioning=0` and or `partition_strategy=wildcard`. File {}",
```
"Give it a try" is too conversational.